cmd/pileup/pileup.go

```diff
@@ -0,0 +1,116 @@
+// Command Group: "SAM Tools"
+
+package main
+
+import (
+	"flag"
+	"fmt"
+	"github.com/vertgenlab/gonomics/dna"
+	"github.com/vertgenlab/gonomics/exception"
+	"github.com/vertgenlab/gonomics/fileio"
+	"github.com/vertgenlab/gonomics/sam"
+	"log"
+	"strings"
+)
+
+func usage() {
+	fmt.Print(
+		"pileup - Count bases from sequencing data\n\n" +
+			"pileup prints to stdout by default, but can be redirected with the -o flag.\n" +
+			"The output format is a tab-separated file with columns defined as:\n" +
+			"Chr\tPos\t#A\t#C\t#G\t#T\t#N\t#DEL\tINS_array\n" +
+			"INS_array is an array of observed insertions with format insSeq:readCount\n\n" +
+			"Usage:\n" +
+			"  pileup [options] in.bam\n\n" +
+			"Options:\n")
+	flag.PrintDefaults()
+}
+
+type Settings struct {
+	minDp int
+}
+
+// pileFilters parses the Settings struct to make functions to filter the output piles.
+func pileFilters(s Settings) []func(pile sam.Pile) bool {
+	var filters []func(pile sam.Pile) bool
+
+	if s.minDp > 0 {
+		filters = append(filters, func(pile sam.Pile) bool {
+			var count int
+			for i := range pile.Count {
+				count += pile.Count[i]
+			}
+			for _, i := range pile.InsCount {
+				count += i
+			}
+			return count >= s.minDp
+		})
+	}
+
+	return filters
+}
+
+func pileup(infile string, outfile string, settings Settings) {
+	sBuilder := new(strings.Builder)
+	reads, recycle, header := sam.GoReadToChanRecycle(infile, 1000)
+	sendChan := make(chan sam.Sam) // no buffer to satisfy recycle contract
```

Comment:
The problem was that the send function needs to know when a read is finished with. It seemed the simplest way to ensure the previous read had been processed was to make the channel unbuffered so that it would block until the calling function was ready for the next read (ensuring that the previous was finished).